text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,303 @@
+
+A reference-level explanation is made for each of the items enumerated above.
+
+## 1 - Allow developers of root crates to recompile `core`
+
+### Use Case
+
+For developers working with new targets not yet supported by the Rust project, this feature would allow the compilation of `core` for any target that can be specified as a valid [target json format].
+
+[target json format]: https://rust-lang.github.io/rfcs/0131-target-specification.html
+
+This functionality would be possible even with the use of a stable compiler.
+
+Users of a nightly compiler would be able to set compile time feature flags for `core` through settings made in their `Cargo.toml`.
+
+### Caveats
+
+For users of a stable compiler, it would not be possible to modify the source code contents of `core`, or change any compile time features of `core` from the defaults used when publishing pre-compiled versions of `core`.
+
+The source code used to build `core` would be the same as the compiler used for building the current project.
+
+### User Interaction
+
+When compiling for a non-standard target, users may specify their target using a json file, rather than a pre-defined target.
+
+For example, currently a user may cross-compile by specifying a target known by Rust:
+
+```sh
+cargo build --target thumbv7em-none-eabihf
+```
+
+Users would also be able to specify a json file, by providing a path to the json file to be used.
+
+```sh
+cargo build --target thumbv7em-freertos-eabihf.json
+```
+
+In general, any of the following would prompt Cargo to recompile `core`, rather than use a pre-compiled version:
+
+* A custom target json is used
+* The root crate has modified the feature flags of `core`
+* The root crate has set certain profile settings, such as opt-level, etc.
+* The root crate has specified a `patch.sysroot` (this is defined in a later section)
+
+Users of a stable compiler would not be able to customize `core` outside of these profile settings.
+
+For users of a nightly compiler, compile time features of `core` may be specified using the same syntax used for other crate dependencies. These specified features may include unstable features.
+
+```toml
+[dependencies.core]
+default-features = false
+features = [...]
+```
+
+It is not necessary to explicitly mention the dependency of `core`, unless changes to features are necessary.
+
+Cargo would use the source of `core` located in the user's `SYSROOT` directory. This source code would be obtained in the same was as necessary today, through the use of `rustup component add rust-src`. If this component is missing, Cargo would exit with an error code, and would prompt the user to execute the command specified above.
+
+### Technical Implications
+
+#### Stabilization of JSON target format
+
+As the custom target json files would become part of the stable interface of Cargo. The format used by this JSON file must become stabilized, and further changes must be made in a backwards compatible way to guarantee stability.
```

Comment:
I always wondered why JSON is used for target specifications. I assume for historical reasons?

I know that it would be more work, but how about switching to TOML? It would be consistent with the other configuration files used by Rust (Cargo.toml, Cargo.lock, .cargo/config) and has the big advantage that it supports comments, which seems very useful because some target options are anything but self-documenting. It would also allow grouping the options in sections, e.g. for separating architecture options (data-layout etc) from build options (linker-flavor etc) if we like.